### PR TITLE
Adding test infrastructure to use mocked service

### PR DESCRIPTION
**Pull request by KKhurin**
### Summary of the changes 
This is part 2 of the message order fix. It introduces:
 - Minor changes to ServiceHubDispatcher to allow replacing `WebSocketConnectionContext` with `MockServiceConnectionContext`
 - Very primitive (intentionally, to shorten this PR) `MockService` which is able to read/write to `MockServiceConnectionContext`
 - Introducing new project with "Integration tests"
 - boilerplate code for the integration tests
 - minor fix to ServiceConnectionBase

### Integration tests
These tests lay in between Unit and E2E categories. Unlike unit tests they don't mock SDK classes and therefore allow more realistic tests and scenarios. Mocking the service code allows for very targeted tests which otherwise would be very difficult when running against the actual service code in E2E approach. Message order fix (which this PR is part of) is one such scenario.

### CODE REVIEW

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+               .UseStartup<IntegrationTestStartup>()
+               .UseUrls("http://localhost:8901")
+               .UseKestrel()
+               .Build();
+
+            await host.StartAsync();
+
+            var mockSvc = host.Services.GetRequiredService<ServiceHubDispatcher<TestHub>>() as MockServiceHubDispatcher<TestHub>;
+            Console.WriteLine(mockSvc);
+            bool result = await mockSvc.MockService.CompletedServiceConnectionHandshake.Task;
```
Comment (KKhurin) [inline test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:42]: This is a very simplistic test - it just verifies that SDK made an attempt to connect to our mock service (e.g. all the wiring worked correctly).

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        [Fact]
+        public async Task MockServiceConnectionHandshake()
+        {
+            Console.WriteLine();
+            var host = new WebHostBuilder()
+               .ConfigureServices(services =>{})
+               .ConfigureLogging(logging => logging.AddXunit(_output))
+               .UseStartup<IntegrationTestStartup>()
+               .UseUrls("http://localhost:8901")
```
Comment (KKhurin) [inline test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:34]: Perhaps we should not start a web server ourselves here and find a simpler way to use SDK.

Comment (vicancy) [reply test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:34]: can [TestServer](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.testhost.testserver?view=aspnetcore-3.0) be leveraged?

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        public IntegrationTestSkeleton(ITestOutputHelper output) : base(output)
+        {
+            _output = output;
+        }
+
+        [Fact]
+        public async Task MockServiceConnectionHandshake()
+        {
+            Console.WriteLine();
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:29]: remove console writeline?

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+               .ConfigureLogging(logging => logging.AddXunit(_output))
+               .UseStartup<IntegrationTestStartup>()
+               .UseUrls("http://localhost:8901")
+               .UseKestrel()
+               .Build();
+
+            await host.StartAsync();
+
+            var mockSvc = host.Services.GetRequiredService<ServiceHubDispatcher<TestHub>>() as MockServiceHubDispatcher<TestHub>;
+            Console.WriteLine(mockSvc);
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:41]: remove console write?

File: src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs [hunk 5]
```diff
@@ -17,17 +17,17 @@ internal class ServiceHubDispatcher<THub> where THub : Hub
     {
         private static readonly string Name = $"ServiceHubDispatcher<{typeof(THub).FullName}>";
 
-        private readonly ILoggerFactory _loggerFactory;
+        protected readonly ILoggerFactory _loggerFactory;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs:20]: no need to change it to `protected`? The mock dispatcher can assign it locally if need to

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+using Microsoft.Azure.SignalR.IntegrationTests.Infrastructure;
+
+namespace Microsoft.Azure.SignalR.IntegrationTests
+{
+
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs:16]: nit: remove additional empty line

File: test/Microsoft.Azure.SignalR.IntegrationTests/Infrastructure/MockServiceConnection.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Microsoft.Azure.SignalR.IntegrationTests.Infrastructure
+{
+
+    /// <summary>
+    /// Encapsulates the actual ConnectionContext to facilitate sync up of MockService & SDK connections
+    /// </summary>
+    internal class MockServiceConnection : IServiceConnection
+    {
+        IServiceConnection _serviceConnection;
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/Infrastructure/MockServiceConnection.cs:17]: nit: add `private readonly`

File: test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs (hunk 2 above)
Comment (KKhurin) [reply test/Microsoft.Azure.SignalR.IntegrationTests/BasicIntegrationTest.cs]: 👍 thanks for the idea!

File: test/Microsoft.Azure.SignalR.IntegrationTests/Microsoft.Azure.SignalR.IntegrationTests.csproj [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    <DefineConstants>MULTIFRAMEWORK</DefineConstants>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.Azure.SignalR.Tests.Common\Microsoft.Azure.SignalR.Tests.Common.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.AspNetCore.TestHost" Version="3.0.3" />
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/Microsoft.Azure.SignalR.IntegrationTests.csproj:16]: move Version value to `dependencies.props`

File: test/Microsoft.Azure.SignalR.IntegrationTests/MockService/MockServiceMock.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace Microsoft.Azure.SignalR.IntegrationTests.MockService
+{
+    // This is just a mock of the ServiceMock (overly simplistic to shorten the current PR)
+    public class MockServiceMock : IMockService
+    {
+        public TaskCompletionSource<bool> CompletedServiceConnectionHandshake { get; } = new TaskCompletionSource<bool>();
+        public IDuplexPipe MockServicePipe { get; set; }
+
+        Task _processIncoming;
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/MockService/MockServiceMock.cs:16]: as a convention, we always encourage adding access modifier specifically.

File: test/Microsoft.Azure.SignalR.IntegrationTests/MockService/MockServiceMock.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+        public Task StartAsync()
+        {
+            var servicePro = new ServiceProtocol();
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.IntegrationTests/MockService/MockServiceMock.cs:20]: a static readonly field is enough?

**Pull request merged**